### PR TITLE
DAO-595 Implement and test EVM script factories for RMC

### PR DESCRIPTION
**Pull request by katamarinaki**
Adds three new EVM Script factories:
* AddMEVBoostRelays
* RemoveMEVBoostRelays
* EditMEVBoostRelays

They allow Relay Management Committee (RMC, trusted caller) to manage MAV-Boost relay allowed list through the EasyTrack motions. 

Deployed contracts in docs: https://docs.lido.fi/deployed-contracts/#easy-track-factories-for-mev-boost-relay-allowed-list-management

### CODE REVIEW

File: contracts/EVMScriptFactories/RemoveMEVBoostRelays.sol [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    /// @param _evmScriptCallData Encoded relay URIs: string[]
+    function createEVMScript(
+        address _creator,
+        bytes memory _evmScriptCallData
+    ) external view override onlyTrustedCaller(_creator) returns (bytes memory) {
+        string[] memory decodedCallData = _decodeEVMScriptCallData(_evmScriptCallData);
+
+        uint256 calldataLength = decodedCallData.length;
+
+        // validate that the call data is not empty, we do not need to check for the relay count as it will be checked on URI exists step
```
[the remark this answers is not in the document]
Comment (swissarmytowel) [reply contracts/EVMScriptFactories/RemoveMEVBoostRelays.sol:65]: bad wording, this means that we can skip the check that amount of relays to be removed does not exceed the count of the relays in the list as its also the byproduct of the validation for pre-existing URIs

File: contracts/EVMScriptFactories/AddMEVBoostRelays.sol [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        address _mevBoostRelayAllowedList
+    ) TrustedCaller(_trustedCaller) {
+        mevBoostRelayAllowedList = IMEVBoostRelayAllowedList(_mevBoostRelayAllowedList);
+    }
+
+    // -------------
+    // EXTERNAL METHODS
+    // -------------
+
+    /// @notice Creates EVMScript to add new MEV boost relay to MEV Boost relay allow list
```
[the remark this answers is not in the document]
Comment (swissarmytowel) [reply contracts/EVMScriptFactories/AddMEVBoostRelays.sol:56]: [fixed in dc5a214](https://github.com/lidofinance/easy-track/pull/81/commits/dc5a214f04aeb29ee0628161510b9d69c9053e5c)

File: contracts/EVMScriptFactories/AddMEVBoostRelays.sol [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    /// @param _evmScriptCallData Encoded relays count and new data: IMEVBoostRelayAllowedList.Relay[]
+    function createEVMScript(
+        address _creator,
+        bytes memory _evmScriptCallData
+    ) external view override onlyTrustedCaller(_creator) returns (bytes memory) {
+        IMEVBoostRelayAllowedList.Relay[] memory decodedCallData = _decodeEVMScriptCallData(
+            _evmScriptCallData
+        );
+
+        uint256 calldataLength = decodedCallData.length;
```
[the remark this answers is not in the document]
Comment (swissarmytowel) [reply contracts/EVMScriptFactories/AddMEVBoostRelays.sol:67]: fixed in [dc5a214](https://github.com/lidofinance/easy-track/pull/81/commits/dc5a214f04aeb29ee0628161510b9d69c9053e5c)

File: contracts/EVMScriptFactories/AddMEVBoostRelays.sol [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        return
+            EVMScriptCreator.createEVMScript(
+                address(mevBoostRelayAllowedList),
+                methodIds,
+                encodedCalldata
+            );
+    }
+
+    /// @notice Decodes call data used by createEVMScript method
+    /// @param _evmScriptCallData Encoded relays count and new data: IMEVBoostRelayAllowedList.Relay[]
```
[the remark this answers is not in the document]
Comment (swissarmytowel) [reply contracts/EVMScriptFactories/AddMEVBoostRelays.sol]: fixed in [dc5a214](https://github.com/lidofinance/easy-track/pull/81/commits/dc5a214f04aeb29ee0628161510b9d69c9053e5c)

File: contracts/EVMScriptFactories/AddMEVBoostRelays.sol [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        relays = abi.decode(_evmScriptCallData, (IMEVBoostRelayAllowedList.Relay[]));
+    }
+
+    function _validateRelayURI(
+        string memory _relayInputURI,
+        uint256 _currentIndex,
+        IMEVBoostRelayAllowedList.Relay[] memory _relays
+    ) private view {
+        require(bytes(_relayInputURI).length > 0, ERROR_EMPTY_RELAY_URI);
+        require(_isRelayUriAvailable(_relayInputURI), ERROR_RELAY_URI_ALREADY_EXISTS);
```
[the remark this answers is not in the document]
Comment (swissarmytowel) [reply contracts/EVMScriptFactories/AddMEVBoostRelays.sol]: fixed in [dc5a214](https://github.com/lidofinance/easy-track/pull/81/commits/dc5a214f04aeb29ee0628161510b9d69c9053e5c)

File: contracts/EVMScriptFactories/RemoveMEVBoostRelays.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (swissarmytowel) [reply contracts/EVMScriptFactories/RemoveMEVBoostRelays.sol]: fixed in [dc5a214](https://github.com/lidofinance/easy-track/pull/81/commits/dc5a214f04aeb29ee0628161510b9d69c9053e5c)

File: contracts/EVMScriptFactories/RemoveMEVBoostRelays.sol [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+                ERROR_RELAY_URI_DUPLICATE
+            );
+
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    function _relayURIExists(string memory _uri) private view returns (bool) {
```
[the remark this answers is not in the document]
Comment (swissarmytowel) [reply contracts/EVMScriptFactories/RemoveMEVBoostRelays.sol]: fixed in [dc5a214](https://github.com/lidofinance/easy-track/pull/81/commits/dc5a214f04aeb29ee0628161510b9d69c9053e5c)

File: contracts/EVMScriptFactories/AddMEVBoostRelays.sol [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+        // check for duplicates in the input data array, starting from the current index for efficiency
+        // if a duplicate is found, it will throw an exception
```
[the remark this answers is not in the document]
Comment (swissarmytowel) [reply contracts/EVMScriptFactories/AddMEVBoostRelays.sol]: fixed in [dc5a214](https://github.com/lidofinance/easy-track/pull/81/commits/dc5a214f04aeb29ee0628161510b9d69c9053e5c)

File: contracts/EVMScriptFactories/RemoveMEVBoostRelays.sol [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+/// @author katamarinaki, swissarmytowel
+/// @notice Creates EVMScript to remove a MEV boost relay from the MEV Boost relay allow list
+contract RemoveMEVBoostRelays is TrustedCaller, IEVMScriptFactory {
+    // -------------
+    // ERRORS
+    // -------------
+
+    string private constant ERROR_EMPTY_CALLDATA = "EMPTY_CALLDATA";
+    string private constant ERROR_REMOVING_MORE_RELAYS_THAN_AVAILABLE =
```
[the remark this answers is not in the document]
Comment (swissarmytowel) [reply contracts/EVMScriptFactories/RemoveMEVBoostRelays.sol]: fixed in [dc5a214](https://github.com/lidofinance/easy-track/pull/81/commits/dc5a214f04aeb29ee0628161510b9d69c9053e5c)

File: contracts/EVMScriptFactories/EditMEVBoostRelays.sol [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+/// @notice Creates EVMScript to edit MEV boost relays in the MEV Boost Relay allow list by the URI
+contract EditMEVBoostRelays is TrustedCaller, IEVMScriptFactory {
+    // -------------
+    // ERRORS
+    // -------------
+
+    string private constant ERROR_EMPTY_CALLDATA = "EMPTY_CALLDATA";
+    string private constant ERROR_EMPTY_RELAY_URI = "EMPTY_RELAY_URI";
+    string private constant ERROR_RELAY_NOT_FOUND = "RELAY_NOT_FOUND";
+    string private constant ERROR_RELAY_URI_DUPLICATE = "DUPLICATE_RELAY_URI";
```
[the remark this answers is not in the document]
Comment (swissarmytowel) [reply contracts/EVMScriptFactories/EditMEVBoostRelays.sol]: fixed in [dc5a214](https://github.com/lidofinance/easy-track/pull/81/commits/dc5a214f04aeb29ee0628161510b9d69c9053e5c)

File: tests/conftest.py [hunk 10]
```diff
@@ -377,6 +377,14 @@ def staking_router(lido_contracts):
 def locator(lido_contracts):
     return lido_contracts.locator
 
+@pytest.fixture(scope="module")
+def mev_boost_relay_allowed_list(lido_contracts, owner):
```
[the remark this answers is not in the document]
Comment (swissarmytowel) [reply tests/conftest.py:391]: fixed in [3197fa9](https://github.com/lidofinance/easy-track/pull/81/commits/3197fa9cbcc3813f397f4637b1dc50d67da27385)

**Review by Psirex — COMMENTED** [review]
LGTM👍🏻

- No major comments on the contracts from my side—mostly nitpicks, which can be treated as recommendations
- Some comments on the tests and scripts that are worth reviewing 

Overall, great work and solid test coverage!

File: tests/evm_script_factories/test_edit_mev_boost_relays.py [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+        edit_mev_boost_relays_factory.createEVMScript(
+            owner,
+            create_calldata([mev_boost_relay_test_config["relays"][0]]),
+        )
+
+
+def test_cannot_edit_relay_not_in_allow_list_with_multiple_relays(
+    owner, edit_mev_boost_relays_factory, mev_boost_relay_allowed_list, mev_boost_relay_test_config
+):
+    "Must revert with message 'RELAY_NOT_FOUND' when relay is not in the allow list"
```
[the remark this answers is not in the document]
Comment (swissarmytowel) [reply tests/evm_script_factories/test_edit_mev_boost_relays.py:285]: Repurposed the test above `test_cannot_edit_more_relays_than_present`, it makes more sense to be i this context

File: tests/evm_script_factories/test_edit_mev_boost_relays.py [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+            create_calldata(
+                [
+                    (f"uri{i}", "o" * (mev_boost_relay_test_config["max_string_length"] + 1), True, f"description{i}")
+                    for i in range(get_max_relay_count(mev_boost_relay_allowed_list, mev_boost_relay_test_config))
+                ]
+            ),
+        )
+
+
+def test_cannot_edit_relay_with_description_over_max_string_length(
```
[the remark this answers is not in the document]
Comment (swissarmytowel) [reply tests/evm_script_factories/test_edit_mev_boost_relays.py:313]: added test case with URI exceeding string limit, and empty URI revert test is present above

File: tests/evm_script_factories/test_remove_mev_boost_relays.py [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+        )
+
+    calldata = create_calldata(inputs)
+    evm_script = remove_mev_boost_relays_factory.createEVMScript(owner, calldata)
+
+    expected_evm_script = encode_call_script(direct_allow_list_calldata)
+    assert evm_script == expected_evm_script
+
+
+def test_remove_max_num_relays(
```
[the remark this answers is not in the document]
Comment (swissarmytowel) [reply tests/evm_script_factories/test_remove_mev_boost_relays.py:128]: it is duplicating checks a little since with the new stub allowed list this test essentially does that (adds max and removes max, clearing the list), but for semantic distinction added a similar test case below with just a few relays to avoid edge case like here

File: tests/integration/conftest.py [hunk 14]
```diff
@@ -316,6 +330,144 @@ def top_up_allowed_recipients_evm_script_factory(
...
+    EditMEVBoostRelays,
+    rmc_factories_multisig,
+    easy_track,
+    lido_contracts,
+    mev_boost_relay_allowed_list,
+    deployer,
+):
+    evm_script_factory = (
+        EditMEVBoostRelays.at(deployed_artifact["EditMEVBoostRelays"]["address"])
+        if "EditMEVBoostRelays" in deployed_artifact
```
Comment (Psirex) [inline tests/integration/conftest.py:437]: should be declared as a parameter of the fixture method before usage:
```python
def edit_mev_boost_relays_evm_script_factory(
    EditMEVBoostRelays,
    rmc_factories_multisig,
    easy_track,
    lido_contracts,
    mev_boost_relay_allowed_list,
    deployer,
    deployed_artifact,
):
```
